source-mongodb/migrate.go

```diff
@@ -0,0 +1,333 @@
+
+		out.DatabaseResumeTokens[db] = tok
+
+		return out, nil, nil
+	}
+
+	// Returning a non-nil token means there is a global resume token that needs to be resolved into
+	// database-level tokens.
+	return out, &tok, nil
+}
+
+// populateDatabaseResumeTokens uses a global resume token for a change stream and populates the
+// capture state's database-specific change stream state from that. It may emit documents, although
+// it will not emit any checkpoints. The caller should emit a full state checkpoint after this
+// function returns to persist the documents and updated checkpoint.
+func (c *capture) populateDatabaseResumeTokens(
+	ctx context.Context,
+	globalToken bson.Raw,
+) error {
+	// Make a set of the database names we need to populate.
+	dbs := make(map[string]struct{})
+	for _, b := range c.collectionBindingIndex {
+		db := b.resource.Database
+		dbs[db] = struct{}{}
+
+		if _, ok := c.state.DatabaseResumeTokens[db]; ok {
+			// Sanity check; this should never happen.
+			return fmt.Errorf("application error: populateDatabaseResumeTokens had state already populated for database %q", db)
+		}
+	}
+
+	log.WithFields(log.Fields{
+		"databases":   dbs,
+		"globalToken": globalToken,
+	}).Info("populating database resume tokens")
+
+	databasePollResults := make(map[string]changeEvent)
+
+	// Poll the global stream. If there's documents available we only need one.
+	opts := options.ChangeStream().SetFullDocument(options.UpdateLookup).SetResumeAfter(globalToken).SetBatchSize(1)
+	cs, err := c.client.Watch(ctx, pl, opts)
+	if err != nil {
+		return fmt.Errorf("opening global change stream: %w", err)
+	}
+	defer cs.Close(ctx)
+
+	gotEvent := false
+	var ev changeEvent
+	var newToken bson.Raw
+
+	for {
+		if cs.TryNext(ctx) {
+			gotEvent = true
+			if err := cs.Decode(&ev); err != nil {
+				return fmt.Errorf("change stream decoding document: %w", err)
+			}
+
+			log.WithFields(log.Fields{
+				"token":       cs.ResumeToken(),
+				"id":          ev.DocumentKey.Id,
+				"ns":          ev.Ns,
+				"clusterTime": ev.ClusterTime,
+			}).Info("v2 state migration got document from global change stream")
+		}
+		if err := cs.Err(); err != nil {
+			return fmt.Errorf("global change stream error: %w", err)
+		}
+
+		newToken = cs.ResumeToken()
+
+		// In rare circumstances we will get the exact same token back as a post-batch resume token,
+		// if the provided global token is very recent. This is probably only really possible in
+		// contrived tests, but we'll handle it anyway by re-polling until we either get an event
+		// from the change stream or a different post-batch resume token.
+		if newToken.String() != globalToken.String() {
+			break
+		}
+
+		if gotEvent {
+			return fmt.Errorf("application error: got an event but resume token didn't change")
+		}
+
+		log.WithFields(log.Fields{
+			"globalToken": globalToken,
+			"newToken":    newToken,
+		}).Info("v2 state migration will re-poll since resume token didn't change")
```

Comment:
Am I right that if there is no event, `newToken.String() != globalToken.String()` is guaranteed to be true, so if we have passed that condition, we definitely have an event, and in that case this log will never be printed since it's after a guaranteed `break`

Reply:
Practically speaking I think that's true, but I was able to produce scenarios in test setups where a post-batch resume token was the same as the global token (which will always be associated with an event), which I think has to do with how MongoDB advances its global operation time. What we're doing with this re-poll is waiting for the global operation time to advance enough where we get a post-batch resume token that is either not associated with an event and can be used as-is for database change streams, or a token for a _new_ event where we know the optime of the event.